llama_index/llms/bedrock.py

```diff
@@ -0,0 +1,223 @@
+        max_tokens_key = bedrock_model_to_param_name(
+            model=self.model, param_name="max_tokens"
+        )
+        base_kwargs = {"temperature": self.temperature, max_tokens_key: self.max_tokens}
+        return {
+            **base_kwargs,
+            **self.additional_kwargs,
+        }
+
+    def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        return {
+            **self._model_kwargs,
+            **kwargs,
+        }
+
+    @llm_completion_callback()
+    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        provider = self.model.split(".")[0]
+        request_body = get_request_body(provider, prompt, all_kwargs)
+        request_body_str = json.dumps(request_body)
+        response = completion_with_retry(
+            client=self._client,
+            model=self.model,
+            request_body=request_body_str,
+            max_retries=self.max_retries,
+            **all_kwargs,
+        )["body"].read()
+        response = json.loads(response)
+        return CompletionResponse(
+            text=get_text_from_response(provider, response), raw=response
+        )
+
+    @llm_completion_callback()
+    def stream_complete(self, prompt: str, **kwargs: Any) -> CompletionResponseGen:
+        if self.model in BEDROCK_FOUNDATION_LLMS and self.model not in STREAMING_MODELS:
+            raise ValueError(f"Model {self.model} does not support streaming")
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        provider = self.model.split(".")[0]
+        request_body = get_request_body(provider, prompt, all_kwargs)
+        request_body_str = json.dumps(request_body)
+        response = completion_with_retry(
+            client=self._client,
+            model=self.model,
+            request_body=request_body_str,
+            max_retries=self.max_retries,
+            stream=True,
+            **all_kwargs,
+        )["body"]
+
+        def gen() -> CompletionResponseGen:
+            content = ""
+            for r in response:
+                r = json.loads(r["chunk"]["bytes"])
+                content_delta = get_text_from_response(provider, r, stream=True)
+                content += content_delta
+                yield CompletionResponse(text=content, delta=content_delta, raw=r)
+
+        return gen()
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        chat_fn = completion_to_chat_decorator(self.complete)
+        return chat_fn(messages, **kwargs)
+
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        stream_chat_fn = stream_completion_to_chat_decorator(self.stream_complete)
+        return stream_chat_fn(messages, **kwargs)
+
+    async def achat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponse:
+        raise NotImplementedError
```

Comment:
thanks @neil-botelho for the contribution. 
I am wondering do you have any plans to add those callback functions?
thanks

Reply:
By callback functions do you mean the async chat and completion endpoints? As far as I can tell, boto3 doesn't support async so we can't make async calls to bedrock. If I'm mistaken or if bedrock adds support to make async calls I'd be happy to update my pr or add a new pr to add those endpoints
References:
- [this stackoverflow article](https://stackoverflow.com/a/72093772)
- [The boto3 bedrock-runtime docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/bedrock-runtime/client/invoke_model.html) lacking an async parameter

Reply:
I did notice the llama_index.llms.CustomLLM is a subclass of LLM that implements the async chat and completion endpoints by just calling the sync chat and completion endpoints from an async function. As far as I can tell, the call to the LLM will still be blocking in this case. Hence I didn't go for that approach.

Reply:
Yep that makes sense, thanks for the explanation!